### ISSUE TITLE
Implementing Rollout in Obs space

### ISSUE THREAD

**Issue by mlaurako** [NOAA-EMC/ocelot#39]
The GraphDOP paper mentions that an optional rollout in observational space can be implemented. However, the paper doesn't clearly specify whether GraphDOP was trained solely on latent rollout or using a mixed approach (both latent and observational rollouts). It would be valuable to run a sensitivity test comparing both methods to evaluate how each rollout type affects model performance.

<img width="1222" alt="Image" src="https://github.com/user-attachments/assets/9b390e39-fd62-4aef-89ac-8a857672ac1a" />

**Comment by mlaurako** [comment]
<html><head></head><body><h1> #115  Milestone 3 — Autoregressive (Observation-Space) Rollout: Implementation Plan</h1>
<h2>Background and Terminology</h2>
<p>OCELOT currently implements <strong>latent rollout</strong>, which is already working.
True AR training requires <strong>observation-space rollout</strong>, which chains multiple
12h prediction windows end-to-end. These are two distinct concepts:</p>

Name | Scope | Status
 -- | -- | --
 Latent rollout | 4 × 3h processor steps inside one 12h window | Working
 Obs-space rollout (AR) | N × 12h windows chained end-to-end | Previous Started but Halted


<p>This document covers AR only.</p>
<hr>
<h2>How AR Works in OCELOT</h2>
<p>Following the GraphDOP approach, AR rollout feeds predictions back as inputs for the
next window — in both training and inference modes:</p>
<pre><code>Window 1:  real obs [t-12h, t)       → encode → process → decode → predictions [t, t+12h)
Window 2:  predictions [t, t+12h)    → encode → process → decode → predictions [t+12, t+24h)
Window 3:  predictions [t+12, t+24h) → encode → process → decode → predictions [t+24, t+36h)
</code></pre>
<p>This differs from GraphCast, which uses two consecutive real states as input
(<code>T1+T2 → T'3</code>, <code>T2+T'3 → T'4</code>, ...) for implicit velocity information.
OCELOT uses a single window per step (<code>T1 → T'2</code>, <code>T'2 → T'3</code>...), consistent
with the observation-space (non-gridded) architecture.</p>
<p><strong>Open question — use real observations during training rollout?</strong>
An alternative training-only strategy grounds each encoder step with real observations
rather than predicted ones:</p>
<pre><code>Window 1:  real obs [t-12h, t)   → encode → process → decode → predictions [t, t+12h)
Window 2:  real obs [t,  t+12h)  → encode → process → decode → predictions [t+12, t+24h)
Window 3:  real obs [t+12, t+24h) → ...
</code></pre>
<p>This is physically analogous to data assimilation and suppresses error accumulation
during early training. The tradeoff is a train/inference mismatch: the model never
sees synthetic inputs during training, which may hurt inference-time AR stability.
GraphDOP uses predicted observations in both modes — this keeps training and inference
consistent and is the default approach here.</p>
<hr>
<h2>What the Processor Already Provides for Free</h2>
<p>The <code>SlidingWindowTransformerProcessor</code> (and its hierarchical variant) was designed
with AR in mind. Its <code>deque</code>-based cache carries temporal context across steps
automatically, and <code>reset()</code> / <code>warm_start()</code> already handle the boundary conditions.</p>
<p><strong>Within one 12h window (current behaviour):</strong></p>
<pre><code>reset()           → cache = []
forward(mesh_t0)  → cache = [t0]
forward(mesh_t1)  → cache = [t0, t1]
forward(mesh_t2)  → cache = [t0, t1, t2]
forward(mesh_t3)  → cache = [t0, t1, t2, t3]   ← window full (maxlen=4)
</code></pre>
<p><strong>Across AR windows (desired behaviour — no code changes to processor needed):</strong></p>
<pre><code># Window 1 — reset at start of sequence only
reset()
forward(mesh_t0)  → cache = [t0]
...
forward(mesh_t3)  → cache = [t0, t1, t2, t3]

#### Window 2 — NO reset(); encode predicted obs for this window
forward(mesh_t4)  → cache = [t1, t2, t3, t4]   ← t0 drops off automatically
forward(mesh_t5)  → cache = [t2, t3, t4, t5]
...
</code></pre>
<p>The <code>deque(maxlen=4)</code> handles oldest-drop-off automatically. <strong>The processor itself
requires zero changes for AR.</strong></p>
<p><code>warm_start()</code> is available for checkpoint resume: pre-populate the cache with the
last few mesh states from a previous sequence (detached, no gradient) before
continuing. This is equivalent to truncated BPTT warmup.</p>
<hr>
<h2>Loss Computation for Long AR Rollouts</h2>
<p>With full BPTT, gradients flow back through all N windows. Peak GPU memory scales
linearly with N. Two practical strategies:</p>
<p><strong>Option A — Full loss across all windows (default for short AR):</strong>
All N windows contribute equally to the loss. Gradient flows through all steps.
Recommended while N is small (2–3 windows).</p>
<p><strong>Option B — Loss on last K windows only (truncated, for longer AR):</strong>
Only the final K steps contribute to the loss; earlier steps are run with
<code>torch.no_grad()</code> or <code>.detach()</code> on the mesh state before the loss window.
This caps memory at K windows regardless of total sequence length.</p>
<pre><code class="language-python">for window_idx, window_data in enumerate(windows):
    is_first = (window_idx == 0)
    is_in_loss_window = (window_idx &gt;= len(windows) - loss_last_k)
    with torch.set_grad_enabled(is_in_loss_window):
        predictions = self._forward_one_ar_window(window_data, is_first)
    if is_in_loss_window:
        total_loss += self.compute_loss(predictions, targets[window_idx])
</code></pre>
<p><strong>Option C — Gradient checkpointing (keep full loss, reduce memory):</strong>
Recompute activations during the backward pass rather than storing them.
<code>torch.utils.checkpoint</code> applied inside <code>_forward_latent_rollout</code> recovers
roughly 50% of the activation memory with ~30% compute overhead.
Recommended before resorting to truncated loss.</p>
<p>In practice: start with Option A (full loss), add gradient checkpointing (Option C)
when memory becomes a constraint, and only fall back to Option B for very long rollouts.</p>
<hr>
<h2>What Needs to Be Built</h2>
<h3>1. Dataset — <code>ARWindowDataset</code></h3>
<p>The current <code>BinDataset</code> returns a single <code>HeteroData</code> per sample.
For AR, each sample must be a sequence of N+1 consecutive windows.</p>
<p>Each <code>__getitem__(idx)</code> loads and returns <code>[window_idx, ..., window_idx+N]</code> —
a list of <code>HeteroData</code> objects for windows at indices <code>idx</code> through <code>idx+N</code>.
<code>ar_steps=1</code> by default reproduces current single-step behaviour exactly, and
<code>__len__</code> = <code>total_bins - ar_steps</code>.</p>
<p><strong>DDP compatibility:</strong>
No custom sampler is needed. Once each dataset item is already a self-contained
sequence of N+1 windows, <code>DistributedSampler</code> works as usual — it distributes
<em>samples</em> (sequences) across GPUs, not individual bins. Each GPU receives
independent non-overlapping sequences.</p>
<pre><code># ar_steps=3, so each sample = 4 consecutive windows
GPU 0: sample_0 = [bin0, bin1, bin2, bin3]
GPU 1: sample_1 = [bin1, bin2, bin3, bin4]
GPU 2: sample_2 = [bin2, bin3, bin4, bin5]
...
</code></pre>
<h3>2. Model Forward Pass — <code>_forward_latent_rollout()</code></h3>
<p>Currently <code>_forward_latent_rollout</code> calls <code>self.swt.reset()</code> unconditionally.
For AR, this must be conditional on whether this is the first window in the sequence:</p>
<pre><code class="language-python">def _forward_latent_rollout(self, data, encoded_features, is_first_ar_window=True):
    if is_first_ar_window:
        self.swt.reset()   # only clear cache at the start of a new AR sequence
    # rest of the loop unchanged
    ...
</code></pre>
<p>No other changes to the forward pass are needed.</p>
<h3>3. Training Loop — AR loop in <code>training_step</code></h3>
<p>A new outer loop over AR windows wrapping the existing single-step forward call.
The first window always encodes real observations; subsequent windows encode the
predictions from the prior window:</p>
<pre><code class="language-python">def training_step(self, batch, batch_idx):
    ar_steps = self.get_current_rollout_steps()   # ← finally gets called
    windows = batch if isinstance(batch, list) else [batch]

    total_loss = 0.0
    current_input = windows[0]   # always real obs for window 1

    for window_idx in range(len(windows) - 1):
        is_first = (window_idx == 0)
        predictions = self._forward_one_ar_window(current_input, is_first_ar_window=is_first)
        total_loss += self.compute_loss(predictions, windows[window_idx + 1])
        # feed predictions back as synthetic input for next window
        current_input = build_synthetic_window(predictions, windows[window_idx + 1])

    return total_loss / (len(windows) - 1)
</code></pre>
<p>Note: <code>build_synthetic_window</code> constructs a <code>HeteroData</code> from predicted values,
with metadata (valid-time, solar time, scan angles, pressure levels) carried
forward from the real target window's node metadata.</p>
<h3>4. Callback — <code>RolloutCurriculumCallback</code></h3>
<p><code>get_current_rollout_steps()</code> exists in <code>gnn_model.py</code> with the full GraphCast-style
curriculum schedule (graphcast / linear / step / fixed) but is never called anywhere.
<code>RolloutCurriculumCallback</code> does not yet exist in <code>callbacks.py</code>.</p>
<pre><code class="language-python">class RolloutCurriculumCallback(pl.Callback):
    def on_train_epoch_end(self, trainer, pl_module):
        new_ar_steps = pl_module.get_current_rollout_steps()
        trainer.datamodule.set_ar_steps(new_ar_steps)
        if trainer.is_global_zero:
            print(f"[RolloutCurriculum] epoch={pl_module.current_epoch} → ar_steps={new_ar_steps}")
</code></pre>
<p><code>datamodule.set_ar_steps(n)</code> (analogous to the existing <code>set_train_window()</code>) triggers
a dataset rebuild with sequence length <code>n+1</code> for the next epoch dataloader reload.</p>
<hr>
<h2>Inference Mode</h2>
<p>AR inference follows the same loop as training: Window 1 encodes real observations;
all subsequent windows encode predictions from the previous step. The key additional
concern vs. training is metadata: target valid-time, local solar time, scan angles,
and pressure levels must be synthesized or carried forward for each predicted window.
The existing <code>_decode_one_step_to_mesh()</code> infrastructure supports this.</p>
<p>Inference AR implementation should follow and be validated against training AR.</p>
<hr>
<h2>Implementation Checklist</h2>
<ul>
<li>[ ] <strong><code>ARWindowDataset</code></strong> — <code>__getitem__</code> returns list of N+1 consecutive <code>HeteroData</code>; <code>ar_steps=1</code> default</li>
<li>[ ] <strong><code>datamodule.set_ar_steps(n)</code></strong> — triggers dataset rebuild; analogous to <code>set_train_window()</code></li>
<li>[ ] <strong><code>is_first_ar_window</code> flag</strong> in <code>_forward_latent_rollout()</code> — suppress <code>reset()</code> between windows</li>
<li>[ ] <strong><code>build_synthetic_window()</code></strong> — construct <code>HeteroData</code> from predictions + metadata from target window</li>
<li>[ ] <strong>AR outer loop in <code>training_step</code></strong> — iterate windows, feed predictions forward, accumulate loss</li>
<li>[ ] <strong><code>RolloutCurriculumCallback</code></strong> in <code>callbacks.py</code> — wire <code>get_current_rollout_steps()</code> → <code>set_ar_steps()</code></li>
<li>[ ] <strong>Smoke test at <code>ar_steps=1</code></strong> — verify refactored code reproduces current single-step loss exactly</li>
<li>[ ] <strong>Gradient checkpointing</strong> inside <code>_forward_latent_rollout</code> — enable before scaling past ar_steps=2</li>
<li>[ ] <strong>Inference AR loop</strong> — synthetic window construction with metadata carry-forward</li>
</ul></body></html>

**Comment by mlaurako** [comment]
### Testing Latent Rollout
The test was run with the following parameters:
```
--data_window_hours 24 
--latent_step_hours 3
```
Using a `3`-year training setup and running for `20` epochs to check functionality. The mesh output of surface temperature from epoch 20 is shown below:
<img width="593" height="424" alt="Image" src="https://github.com/user-attachments/assets/b2120b3c-5ca1-44b9-95db-b6e2867f1be3" />
